shared/src/bad_token/trace_call.rs

```diff
@@ -209,76 +211,146 @@ impl TraceCallDetector {
         // 6
         let tx = instance.balance_of(recipient).m.tx;
         requests.push(call_request(None, token, tx));
-
+        // arbitrary recipient -> pool
         // 7
+        let tx = instance.balance_of(pool).m.tx;
+        requests.push(call_request(None, token, tx));
+        // 8
+        let tx = instance.transfer(pool, amount).tx;
+        requests.push(call_request(Some(recipient), token, tx));
+        // 9
+        let tx = instance.balance_of(pool).m.tx;
+        requests.push(call_request(None, token, tx));
+
+        // 10
         let tx = instance.approve(recipient, U256::MAX).tx;
         requests.push(call_request(Some(self.settlement_contract), token, tx));
 
         requests
     }
 
     fn handle_response(traces: &[BlockTrace], amount: U256) -> Result<TokenQuality> {
-        ensure!(traces.len() == 8, "unexpected number of traces");
+        ensure!(traces.len() == 11, "unexpected number of traces");
 
-        let gas_in = match ensure_transaction_ok_and_get_gas(&traces[1])? {
+        let gas_to_settlement = match ensure_transaction_ok_and_get_gas(&traces[1])? {
             Ok(gas) => gas,
-            Err(reason) => return Ok(TokenQuality::bad(reason)),
+            Err(reason) => {
+                return Ok(TokenQuality::bad(format!(
+                    "failed transaction to settlement: {}",
+                    reason
+                )))
+            }
+        };
+        let gas_to_recipient = match ensure_transaction_ok_and_get_gas(&traces[4])? {
+            Ok(gas) => gas,
+            Err(reason) => {
+                return Ok(TokenQuality::bad(format!(
+                    "failed transaction to settlement: {}",
+                    reason
+                )))
+            }
         };
-        let gas_out = match ensure_transaction_ok_and_get_gas(&traces[4])? {
+        let gas_to_pool = match ensure_transaction_ok_and_get_gas(&traces[8])? {
             Ok(gas) => gas,
-            Err(reason) => return Ok(TokenQuality::bad(reason)),
+            Err(reason) => {
+                return Ok(TokenQuality::bad(format!(
+                    "failed transaction to pool: {}",
+                    reason
+                )))
+            }
         };
 
-        let balance_before_in = match decode_u256(&traces[0]) {
+        let bal_settlement_initial = match decode_u256(&traces[0]) {
             Ok(balance) => balance,
             Err(_) => return Ok(TokenQuality::bad("can't decode initial settlement balance")),
         };
-        let balance_after_in = match decode_u256(&traces[2]) {
+        let bal_settlement_after_transfer_from_pool = match decode_u256(&traces[2]) {
```

Comment:
These one use the short term bal but the latter ones use balance. Maybe we can be consistent here. 

Also, given that there are so many balances, maybe we could make a balance struct with all these things are fields.